burn-wgpu/src/kernel/binary_elemwise.rs

```diff
@@ -0,0 +1,159 @@
+use super::{KernelGenerator, KernelSettings};
+use crate::{context::WorkGroup, element::WGPUElement, kernel_wgsl, tensor::WGPUTensor};
+use burn_tensor::Shape;
+use num_traits::ToPrimitive;
+use std::sync::Arc;
+
+kernel_wgsl!(BinaryElemwiseRaw, "../template/binary_elemwise.wgsl");
+kernel_wgsl!(
+    BinaryElemwiseInplaceRaw,
+    "../template/binary_elemwise_inplace.wgsl"
+);
+
+#[macro_export]
+macro_rules! binary_elemwise {
+    (
+        $struct:ident,
+        $ops:expr
+    ) => {
+        pub struct $struct;
+
+        impl $crate::kernel::KernelGenerator for $struct {
+            type Source = String;
+
+            fn generate() -> Self::Source {
+                let source = $crate::kernel::BinaryElemwiseRaw::generate().to_string();
+                let body = format!(
+                    "output[global_id.x] = lhs[index_lhs] {} rhs[index_rhs]",
+                    $ops
+                );
+                source.replace("BODY", &body)
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! binary_elemwise_inplace {
+    (
+        $struct:ident,
+        $ops:expr
+    ) => {
+        pub struct $struct;
+
+        impl $crate::kernel::KernelGenerator for $struct {
+            type Source = String;
+
+            fn generate() -> Self::Source {
+                let source = $crate::kernel::BinaryElemwiseInplaceRaw::generate().to_string();
+                let body = format!(
+                    "lhs[global_id.x] = lhs[global_id.x] {} rhs[index_rhs];",
+                    $ops
+                );
+                source.replace("BODY", &body)
+            }
+        }
+    };
+}
+
+pub fn binary_elemwise<K: KernelGenerator, E: WGPUElement, const D: usize>(
+    lhs: WGPUTensor<E, D>,
+    rhs: WGPUTensor<E, D>,
+) -> WGPUTensor<E, D> {
+    lhs.assert_is_on_save_device(&rhs);
+
+    let mut shape_out = [0; D];
+    lhs.shape
+        .dims
+        .iter()
+        .zip(rhs.shape.dims.iter())
+        .enumerate()
+        .for_each(|(index, (dim_lhs, dim_rhs))| {
+            shape_out[index] = usize::max(*dim_lhs, *dim_rhs);
+        });
+
+    let shape_out = Shape::new(shape_out);
+
+    let buffer = lhs
+        .context
+        .create_buffer(shape_out.num_elements() * core::mem::size_of::<E>());
+    let output = WGPUTensor::new(lhs.context.clone(), shape_out, Arc::new(buffer));
+
+    let kernel = lhs
+        .context
+        .compile::<KernelSettings<K, E, i32, 256, 1, 1>>();
```

Comment:
Probably we want to document this 256 magic number into a constant. Not sure if this will need to change again.

Reply:
This is the workgroup size. For now it is hardcoded, but I would like execute multiple kernels with different parameters to see which one performs the best (at runtime). This depends on the tensor shape and the actual GPU.